plugins/colorbutton/plugin.js

```diff
@@ -331,45 +382,277 @@ CKEDITOR.plugins.add( 'colorbutton', {
+							'<tr>' +
+								'<td colspan="' + colorHistoryRow.colorsPerRow + '" align="center">' +
+									'<span class="cke_colorbox" id="' + colorBoxId + '"></span>' + lang.auto +
+								'</td>' +
+							'</tr>' +
+						'</table>' +
+					'</a>';
+			}
+
+			function generateMoreColorsButtonHtml() {
+				return '</tr>' +
+					'<tr>' +
+						'<td colspan="' + colorHistoryRow.colorsPerRow + '" align="center">' +
+							'<a class="cke_colormore" _cke_focus=1 hidefocus=true' +
+								' title="' + lang.more + '"' +
+								' draggable="false"' +
+								' ondragstart="return false;"' + // Draggable attribute is buggy on Firefox.
+								' onclick="CKEDITOR.tools.callFunction(' + clickFn + ',\'?\',\'' + type + '\');return false;"' +
+								' href="javascript:void(\'' + lang.more + '\')"' + ' role="option" aria-posinset="' + total +
+								'" aria-setsize="' + total + '">' + lang.more +
+							'</a>' +
+						'</td>'; // </tr> is later in the code.
+			}
+
 			function setColor( color ) {
 				var colorStyle = color && new CKEDITOR.style( colorStyleTemplate, { color: color } );
 
 				editor.execCommand( commandName, { newStyle: colorStyle } );
 			}
 		}
 
+		// This function is called on the first panel opening.
+		function fillColorHistory( options ) {
+			if ( !colorHistoryRow.rowLimit ) {
+				return;
+			}
+
+			var colorSpans = editor.getStyledSpans( options.cssProperty ),
+				htmlColorsList = CKEDITOR.tools.style.parse._colors,
+				colorNames = editor.lang.colorbutton.colors,
+				colorOccurrences,
+				sortedColors;
+
+			if ( !colorSpans.length ) {
+				return;
+			}
+
+			colorOccurrences = extractColorInfo( colorSpans, options.cssProperty, htmlColorsList );
+
+			if ( CKEDITOR.tools.isEmpty( colorOccurrences ) ) {
+				return;
+			}
+
+			sortedColors = sortByOccurrencesAscending( colorOccurrences, 'colorCode' );
+
+			sortedColors.splice( colorHistoryRow.capacity() );
+
+			addLabels( sortedColors, colorNames );
+
+			createColorBoxes( {
+				colorArray: sortedColors,
+				clickFn: options.clickFn,
+				setSize: sortedColors.length,
+				rowSize: colorHistoryRow.colorsPerRow,
+				row: options.colorHistoryRow
+			} );
+
+			options.colorHistorySeparator.show();
+		}
+
+		function extractColorInfo( spans, cssProperty, colorList ) {
+			var counterObject = {};
+
+			CKEDITOR.tools.array.forEach( spans, function( span ) {
+				var spanColor = getHexCode( span, colorList, cssProperty );
+
+				countOccurrences( spanColor, counterObject );
+			} );
+
+			return counterObject;
+		}
+
+		function getHexCode( span, list, cssProperty ) {
+			var color = span.getStyle( cssProperty );
+
+			return color in list ? list[ color ].substr( 1 ) : normalizeColor( span.getComputedStyle( cssProperty ) ).toUpperCase();
+		}
+
+		function countOccurrences( property, object ) {
+			if ( property in object ) {
+				object[ property ] += 1;
+			} else {
+				object[ property ] = 1;
+			}
+		}
```

Comment:
AFAIK we talked about this function :thinking: nevertheless, it could be simplified by using `reduce` function instead:
```javascript
function extractColorInfo( spans, cssProperty, colorList ) {
	return CKEDITOR.tools.array.reduce( spans, function( counterObject, span ) {
		var hexCode = getHexCode( span, colorList, cssProperty );

		counterObject[ hexCode ] = counterObject[ hexCode ] || 0;
		counterObject[ hexCode ] += 1;
		return counterObject;

	}, {} );
}

```

Reply:
Refactored as suggested.